tools/cldr-code/src/test/java/org/unicode/cldr/api/CldrDataProcessorTest.java

```diff
@@ -0,0 +1,135 @@
+    private static final class CurrencyData {
+        final String key;
+        String name = "";
+        String symbol = "";
+
+        CurrencyData(String key) {
+            this.key = key;
+        }
+
+        CurrencyData(String key, String name, String symbol) {
+            this.key = key;
+            this.name = name;
+            this.symbol = symbol;
+        }
+
+        @Override public boolean equals(Object o) {
+            if (o instanceof CurrencyData) {
+                CurrencyData that = (CurrencyData) o;
+                return key.equals(that.key) && name.equals(that.name) && symbol.equals(that.symbol);
+            }
+            return false;
+        }
+
+        @Override public int hashCode() {
+            return Objects.hash(key, name, symbol);
+        }
+
+        @Override public String toString() {
+            return String.format("CurrencyData{name=%s, symbol='%s'}", name, symbol);
+        }
+    }
+
+    // For collecting processed values.
+    private static final class State {
+        ImmutableMap<String, String> names = ImmutableMap.of();
+        ImmutableMap<String, CurrencyData> currencies = ImmutableMap.of();
+
+        void setNames(Map<String, String> map) {
+            names = ImmutableMap.copyOf(map);
+        }
+
+        void setCurrencies(Map<String, CurrencyData> map) {
+            currencies = ImmutableMap.copyOf(map);
+        }
+    }
+
+    private static final CldrDataProcessor<State> VISITOR = createTestVisitor();
+
+    private static CldrDataProcessor<State> createTestVisitor() {
+        // Note that this is deliberately doing things the "messy" way by creating and then copying
+        // a map. This is to show an extra level of processing in tests. You could just have a
+        // value action which adds the territory to a map in the State object.
+        CldrDataProcessor.Builder<State> builder = CldrDataProcessor.builder();
+        builder
+            .addAction(
+                "//ldml/localeDisplayNames/territories",
+                () -> new LinkedHashMap<String, String>(),
+                State::setNames)
+            .addValueAction(
+                "territory[@type=*]",
+                (map, value) -> map.put(value.getPath().get(TERRITORY_TYPE), value.getValue()));
+
+        // Another convoluted example for testing. This has the same additional level for a map
+        // just so we can show a 3-level processor. In real code this wouldn't look so messy.
+        CldrDataProcessor.SubProcessor<CurrencyData> currencyProcessor = builder
+            .addAction(
+                "//ldml/numbers/currencies",
+                () -> new LinkedHashMap<String, CurrencyData>(),
+                State::setCurrencies)
+            .addAction(
+                "currency[@type=*]",
+                (map, path) -> new CurrencyData(path.get(CURRENCY_TYPE)),
+                (map, data) -> map.put(data.key, data));
```

Comment:
I suggest better names for the lambda arguments reflecting the API rather than the test case.  Isn't the type of `map` as you have it on line 89 (the unused 1st parameter) actually `CurrencyData`? If so then `map` would definitely be misleading.

```suggestion
            .addAction(
                "currency[@type=*]",
                (state, path) -> new CurrencyData(path.get(CURRENCY_TYPE)),
                (state, substate) -> map.put(substate.key, substate));
```

Reply:
I couldn't find any _examples_ where the `BiFunction<T, …> newStateFn` uses the `T` parameter (state). 
`addSubprocessor(String)` uses it internally, to be sure. An idea would be to add a `state.allocateList()` function or something, to show that the `newStateFn` sometimes needs `T`  when allocating a new `S`.  That's minor and definitely contrived.